lib/ruby_lsp/requests/document_highlight.rb

```diff
@@ -0,0 +1,153 @@
+# frozen_string_literal: true
+
+module RubyLsp
+  module Requests
+    # The [document highlight](https://microsoft.github.io/language-server-protocol/specification#textDocument_documentHighlight)
+    # informs the editor all relevant elements of the currently pointed item for highlighting. For example, when
+    # the cursor is on the `F` of the constant `FOO`, the editor should identify other occurences of `FOO`
+    # and highlight them.
+    #
+    # For writable elements like constants or variables, their read/write occurrences should be highlighted differently.
+    # This is achieved by sending different "kind" attributes to the editor (2 for read and 3 for write).
+    #
+    # # Example
+    #
+    # ```ruby
+    # FOO = 1 # should be highlighted as "write"
+    #
+    # def foo
+    #   FOO # should be highlighted as "read"
+    # end
+    # ````
+    class DocumentHighlight < BaseRequest
+      def self.run(document, position)
+        new(document, position).run
+      end
+
+      def initialize(document, position)
+        position = Document::Scanner.new(document.source).find_position(position)
+        @visitor = find(document.tree, position)
+        @highlights = []
+
+        super(document)
+      end
+
+      def run
+        # no visitor means the target is not highlightable
+        return [] unless @visitor
+
+        @visitor.visit(@document.tree)
+        @visitor.highlights
+      end
+
+      private
+
+      def find(node, position)
+        matched =
+          node.child_nodes.compact.bsearch do |child|
+            if (child.location.start_char...child.location.end_char).cover?(position)
+              0
+            else
+              position <=> child.location.start_char
+            end
+          end
+
+        case matched
+        when SyntaxTree::GVar
+          GVarHighlightVisitor.new(matched)
+        when SyntaxTree::Ident
+          IdentHighlightVisitor.new(matched)
+        when SyntaxTree::IVar
+          IVarHighlightVisitor.new(matched)
+        when SyntaxTree::Const
+          ConstHighlightVisitor.new(matched)
+        when SyntaxTree::Node
+          find(matched, position)
+        end
+      end
+
+      class HighlightVisitor < SyntaxTree::Visitor
+        attr_reader :highlights
+
+        def initialize(target)
+          @highlights = []
+          @assign_node_stack = []
+          @target = target
+
+          super()
+        end
+
+        def visit_assign(node)
+          @assign_node_stack << node
+          super
+        ensure
+          @assign_node_stack.pop
+        end
+
+        def visit_var_field(node)
```

Comment:
@kddnewton  Can I say `var_field` always represents a "write" to the variable, while `var_ref` is always "read"? Is there any exception?